File: backend/green_gov_rag/api/schemas/lifecycle.py
"""Pydantic schemas for document lifecycle management API."""

from __future__ import annotations

from datetime import datetime
from typing import Optional

from pydantic import BaseModel

# ---------------------------------------------------------------------------
# Summary / counts
# ---------------------------------------------------------------------------


class LifecycleSummary(BaseModel):
    """Counts of DocumentFiles in each lifecycle state."""

    detect: int = 0
    fetch: int = 0
    chunk: int = 0
    embed: int = 0
    available_for_search: int = 0
    url_dead: int = 0
    mark_superseded: int = 0
    removed_from_search: int = 0
    total_files: int = 0
    total_sources: int = 0
    last_monitoring_run: Optional[datetime] = None


# ---------------------------------------------------------------------------
# Document file entry (for the registry table)
# ---------------------------------------------------------------------------


class LifecycleDocumentEntry(BaseModel):
    """One row in the LGA-grouped document registry table."""

    file_id: str
    source_id: str
    title: str
    jurisdiction: str
    topic: str
    lga_names: list[str]  # from spatial_metadata.lga_names
    applies_to_all_lgas: bool
    file_url: str
    lifecycle_state: str
    lifecycle_transitioned_at: Optional[datetime]
    http_status_code: Optional[int]
    http_last_checked_at: Optional[datetime]
    superseded_by_url: Optional[str]


class LifecycleDocumentListResponse(BaseModel):
    """Paginated response for the document registry table."""

    items: list[LifecycleDocumentEntry]
    total: int
    limit: int
    offset: int


# ---------------------------------------------------------------------------
# LGA-grouped view
# ---------------------------------------------------------------------------


class LGADocumentGroup(BaseModel):
    """All documents for a single LGA (or 'All LGAs' for federal/state)."""

    lga_name: str  # e.g. "Adelaide City Council" or "All LGAs"
    documents: list[LifecycleDocumentEntry]


class LGALifecycleResponse(BaseModel):
    """LGA-grouped document registry response."""

    groups: list[LGADocumentGroup]
    total_lgas: int
    total_files: int


# ---------------------------------------------------------------------------
# Lifecycle event history
# ---------------------------------------------------------------------------


class LifecycleEventEntry(BaseModel):
    """One row from lifecycle_event_log."""

    id: int
    from_state: str
    to_state: str
    triggered_by: str
    http_status: Optional[int]
    run_id: Optional[str]
    details: Optional[dict]
    created_at: datetime


# ---------------------------------------------------------------------------
# Admin actions
# ---------------------------------------------------------------------------


class ReplaceDocumentRequest(BaseModel):
    """Admin provides a replacement URL for a url_dead document."""

    new_url: str  # validated as a URL string


class RegisterDocumentRequest(BaseModel):
    """Register a new document source via the API (no YAML PR needed)."""

    title: str
    source_url: str
    download_urls: list[str]
    jurisdiction: str
    category: str
    topic: str
    region: Optional[str] = None
    esg_metadata: Optional[dict] = None
    spatial_metadata: Optional[dict] = None


class RegisterDocumentResponse(BaseModel):
    """Response after registering a new document."""

    source_id: str
    file_ids: list[str]
    lifecycle_state: str = "detect"
    message: str

File: frontend/src/types/lifecycle.ts
// TypeScript types mirroring backend lifecycle Pydantic schemas
// backend/green_gov_rag/api/schemas/lifecycle.py

export type LifecycleState =
  | 'detect'
  | 'fetch'
  | 'chunk'
  | 'embed'
  | 'available_for_search'
  | 'url_dead'
  | 'mark_superseded'
  | 'removed_from_search';

export interface LifecycleSummary {
  detect: number;
  fetch: number;
  chunk: number;
  embed: number;
  available_for_search: number;
  url_dead: number;
  mark_superseded: number;
  removed_from_search: number;
  total_files: number;
  total_sources: number;
  last_monitoring_run: string | null; // ISO datetime
}

export interface LifecycleDocumentEntry {
  file_id: string;
  source_id: string;
  title: string;
  jurisdiction: string;
  topic: string;
  lga_names: string[];
  applies_to_all_lgas: boolean;
  file_url: string;
  lifecycle_state: LifecycleState;
  lifecycle_transitioned_at: string | null;
  http_status_code: number | null;
  http_last_checked_at: string | null;
  superseded_by_url: string | null;
}

export interface LGADocumentGroup {
  lga_name: string;
  documents: LifecycleDocumentEntry[];
}

export interface LGALifecycleResponse {
  groups: LGADocumentGroup[];
  total_lgas: number;
  total_files: number;
}

export interface LifecycleEventEntry {
  id: number;
  from_state: LifecycleState;
  to_state: LifecycleState;
  triggered_by: string;
  http_status: number | null;
  run_id: string | null;
  details: Record<string, unknown> | null;
  created_at: string; // ISO datetime
}

export interface DocumentHistoryResponse {
  file_id: string;
  events: LifecycleEventEntry[];
}

export interface RegisterDocumentRequest {
  title: string;
  source_url: string;
  download_urls: string[];
  jurisdiction: string;
  category: string;
  topic: string;
  region?: string;
  esg_metadata?: Record<string, unknown>;
  spatial_metadata?: {
    spatial_scope?: string;
    state?: string | null;
    lga_codes?: string[];
    lga_names?: string[];
    applies_to_all_lgas?: boolean;
    applies_to_point?: boolean;
  };
}

export interface RegisterDocumentResponse {
  source_id: string;
  file_ids: string[];
  lifecycle_state: string;
  message: string;
}

export interface ReplaceDocumentResponse {
  status: string;
  superseded_file_id: string;
  new_file_id: string;
  new_url: string;
  message: string;
}
